File: day14/Part2.py
import sys
from math import floor

f = open(sys.argv[1], "r")
map = []
for line in f:
	if line[len(line) - 1] == '\n':
		line = line[:len(line) - 1]
	map.append(list(line))


def countChargesNorth(map):
	y = 0
	x = 0
	charges = 0
	while y < len(map):
		x = 0
		while (x < len(map[y])):
			if (map[y][x] == 'O'):
				charges += len(map) - y
			x = x + 1
		y = y + 1
	return (charges)

def slideRocksDirection(map, direction):
	y = len(map) - 1 if direction == 2 else 0
	while (y >= 0 if direction == 2 else y < len(map)):
		x = len(map[y]) - 1 if direction == 3 else 0
		while (x >= 0 if direction == 3 else x < len(map[y])):
			if (map[y][x] == 'O'):
				if (direction == 0 or direction == 2):
					slideCoord = y
					target = 0 if direction == 0 else len(map) - 1
					incrementor = -1 if direction == 0 else 1
					while (slideCoord != target and map[slideCoord + incrementor][x] == '.'):
						slideCoord += incrementor
					map[y][x] = '.'
					map[slideCoord][x] = 'O'
				else:
					slideCoord = x
					target = 0 if direction == 1 else len(map[y]) - 1
					incrementor = -1 if direction == 1 else 1
					while (slideCoord != target and map[y][slideCoord + incrementor] == '.'):
						slideCoord += incrementor
					map[y][x] = '.'
					map[y][slideCoord] = 'O'
			x += -1 if direction == 3 else 1
		y += -1 if direction == 2 else 1


def rawMap(map):
	newMap = []
	for line in map:
		newMap.append(''.join(line))
	return ("\n".join(newMap))

s = {}
nbCycles = 1000000000
i = 0
print(rawMap(map), "\n")
while (i < nbCycles):
	j = 0
	while (j < 4):
		slideRocksDirection(map, j)
		j += 1
	raw = rawMap(map)
	if (rawMap(map) in s):
		nbCycleForMap = i - s[raw]
		print("new correspondance at ", i)
		while (i + nbCycleForMap < nbCycles):
			i += nbCycleForMap
		s = {}
		print("new cycle : ", i, i / nbCycles * 100)
	else:
		s[raw] = i
	i += 1
print(rawMap(map))
print(countChargesNorth(map))

File: day14/Part2.js
const fs = require('fs');
const file = process.argv.length > 2 ? process.argv[2] : "input.txt"

function countChargesNorth(map)
{
    var y = 0;
    var x;
    var charges = 0;
    while (y < map.length)
    {
        x = 0;
        while (x < map[y].length)
        {
            if (map[y][x] == 'O')
                charges += (map.length - y);
            x++;
        }
        y++;
    }
    return (charges);
}

// direction : 0 north  1 west  2  south  3 east
function slideRocksDirection(map, direction)
{
    var y = direction == 2 ? map.length - 1 : 0;
    var slideCoord;
    var x;
    while ((direction == 2) ? y >= 0 : y < map.length)
    {
        x = direction == 3 ? map[y].length - 1 : 0;
        while (direction == 3 ? x >= 0 : x < map[y].length)
        {
            if (map[y][x] == 'O')
            {
                if (direction == 0 || direction == 2)
                {
                    slideCoord = y;
                    let target = direction == 0 ? 0 : map.length - 1;
                    let incrementor = direction == 0 ? -1 : 1;
                    while (slideCoord != target && map[slideCoord + incrementor][x] == '.')
                        slideCoord += incrementor;
                    map[y] = map[y].substring(0, x) + '.' + map[y].substring(x + 1);
                    map[slideCoord] = map[slideCoord].substring(0, x) + 'O' + map[slideCoord].substring(x + 1);
                }
                else
                {
                    slideCoord = x;
                    let target = direction == 1 ? 0 : map[y].length - 1;
                    let incrementor = direction == 1 ? -1 : 1;
                    while (slideCoord != target && map[y][slideCoord + incrementor] == '.')
                        slideCoord += incrementor;
                    map[y] = map[y].substring(0, x) + '.' + map[y].substring(x + 1);
                    map[y] = map[y].substring(0, slideCoord) + 'O' + map[y].substring(slideCoord + 1);
                }
            }
            x += (direction == 3) ? -1 : 1;
        }
        y += (direction == 2) ? -1 : 1;
    }
}

fs.readFile(file, {encoding: "utf-8"}, (err, res) => {
    if (err)
        console.error(err);
    else
    {
        res = res.split('\n');
        console.log(res.join('\n'), '\n\n');
        let nbCycles = 1000000000;
        let quarter = nbCycles / 10;
        let i, j;
        for (i = 0; i < nbCycles; i++)
        {
            for (j = 0; j < 4; j++)
                slideRocksDirection(res, j);
            if (i % quarter == 0)
                console.log(`${i / nbCycles * 100}%`);
        }
        console.log(res.join('\n'));
        console.log(countChargesNorth(res));
    }
})